torch/csrc/autograd/python_variable.cpp

```diff
@@ -0,0 +1,386 @@
+
+#include <structmember.h>
+
+#include "THP.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/autograd/python_cpp_function.h"
+#include "torch/csrc/cuda/AutoGPU.h"
+#include "torch/csrc/utils/auto_gil.h"
+#include <THPP/tensors/THTensor.hpp>
+
+
+using namespace torch::autograd;
+
+PyObject *THPVariableClass = NULL;
+
+static PyObject* THPVariable_NewWithVar(PyTypeObject* type, std::shared_ptr<Variable> var)
+{
+  PyObject* obj = type->tp_alloc(type, 0);
+  if (obj) {
+    auto v = (THPVariable*) obj;
+    new (&v->cdata) std::shared_ptr<Variable>(std::move(var));
+  }
+  return obj;
+}
+
+PyObject * THPVariable_Wrap(std::shared_ptr<Variable>& var)
+{
+  if (var->pyobj) {
+    Py_INCREF(var->pyobj);
+  } else {
+    var->pyobj = THPVariable_NewWithVar((PyTypeObject *)THPVariableClass, var);
+  }
+  return var->pyobj;
+}
+
+// This function DOES NOT steal a reference to data and creator
+// To create a leaf Variable pass NULL as creator.
+PyObject * THPVariable_New(PyObject *data, PyObject *creator, bool requires_grad, bool is_volatile)
+{
+  THPUtils_assert(THPModule_isTensor(data), "data must be a Tensor");
+  THPUtils_assert(!creator || THPFunction_Check(creator), "creator must be a Function");
+  auto v = std::make_shared<Variable>(torch::createTensor(data), requires_grad, is_volatile);
+  PyObject* obj = THPVariable_NewWithVar((PyTypeObject*)THPVariableClass, v);
+  if (obj) {
+    v->pyobj = obj;
+    v->creator = THPFunction_asFunction((THPFunction*)creator);
+    ((THPVariable*)obj)->data = data;
+    Py_INCREF(data);
+  }
+  return obj;
+}
+
+// This function DOES NOT steal a reference to data
+PyObject * THPVariable_NewVolatile(PyObject *data)
+{
+  return THPVariable_New(data, nullptr, false, true);
+}
+
+static int THPVariable_traverse(THPVariable *self, visitproc visit, void *arg)
+{
+  Py_VISIT(self->data);
+  Py_VISIT(self->backward_hooks);
+  return 0;
+}
+
+static int THPVariable_clear(THPVariable *self)
+{
+  Py_CLEAR(self->data);
+  Py_CLEAR(self->backward_hooks);
+  return 0;
+}
+
+static void THPVariable_dealloc(THPVariable* self)
+{
+  PyObject_GC_UnTrack(self);
+  Py_XDECREF(self->data);
+  Py_XDECREF(self->backward_hooks);
+  self->cdata->pyobj = nullptr;
+  self->cdata.~shared_ptr<Variable>();
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+PyObject *THPVariable_pynew(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  THPObjectPtr _data;
+  PyObject *data = NULL;
+  PyObject *creator = NULL;
+  char is_volatile = 0;
+  char requires_grad = 0;
+
+  const char *accepted_args[] = {"data", "creator", "volatile", "requires_grad", NULL};
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|OObb", (char**)accepted_args,
+      &data, &creator, &is_volatile, &requires_grad))
+    return NULL;
+
+  if (creator == Py_None)
+    creator = NULL;
+
+  if (data == NULL || data == Py_None) {
+    // For legacy serialization code, create an empty tensor temporarily.
```

Comment:
In the new code, I want the following invariants to hold because they simplify NULL pointer checking:

- Every THPVariable has a valid Variable 'cdata'
- Every Variable has a valid Tensor 'data'

This is tricky because previously serialized Variables will cause the following invocations during unpickling:

```
tmp = Variable.__new__()
tmp.__setstate__((data, grad, hooks, requires_grad, is_volatile))
```

To work around this, `Variable.__new__` accepts zero arguments. It sets the data to an empty FloatTensor. The tensor is quickly overwritten by the setstate call during unpickling, so the choice of data type doesn't matter.

To prevent user code from creating an empty `Variable()`, `Variable.__init__` requires the `data` argument. This is OK because unpickling doesn't call `__init__`